### PR TITLE
fix: derive durable memory namespace from session keys

### LINKED ISSUE #20: LibraVDB stays empty in CC direct-runtime sessions and assumes userId the host does not provide
## Summary

LibraVDB can appear fully healthy (`sidecar running`, socket ok, embeddings ready) while storing **zero turns / zero memories** in active Command Center usage.

I traced this in a live OpenClaw install and found **two separate compatibility gaps**:

1. **Direct-runtime sessions do not drive the context-engine ingest lifecycle**
2. **The plugin assumes `userId` is provided by the host context-engine contract, but the current host contract does not provide it**

Those combine to make LibraVDB look installed but effectively inert in normal CC usage.

## Observed behavior

`openclaw memory status` shows:

- Sidecar: `running`
- Turns stored: `0`
- Memories stored: `0`
- Lifecycle hints: `0`
- Gate threshold: `0.35`
- Abstractive model: `ready`
- Embedding profile: `nomic-embed-text-v1.5`

At the same time:

- the active session is a CC chat on **`Runtime: direct`**
- `lossless-claw` is recording conversation history normally
- LibraVDB’s DB remains tiny and empty-looking

## Root cause 1: direct runtime does not invoke the context-engine lifecycle

In OpenClaw core, context-engine resolution/invocation is wired through the embedded Pi runner paths, for example:

- `core/openclaw/src/agents/pi-embedded-runner/run.ts`
- `core/openclaw/src/agents/pi-embedded-runner/compact.ts`

The active CC session I traced was on **direct runtime**, and that path does not appear to call the context-engine ingest lifecycle at all.

Result:

- LibraVDB can be selected as `plugins.slots.contextEngine`
- the daemon can be healthy
- but direct-runtime turns from CC are never ingested

So for normal CC/direct usage, the plugin currently receives no ingest calls.

## Root cause 2: plugin expects `userId`, but host contract does not provide it

In the plugin code, the context engine expects `userId` during bootstrap/ingest/assemble:

- `src/types.ts`
- `src/context-engine.ts`

Examples:

- `ContextBootstrapArgs { sessionId, userId }`
- `ContextIngestArgs { sessionId, userId, message }`
- collections like:
  - `turns:${userId}`
  - `user:${userId}`

But OpenClaw’s host context-engine contract does **not** provide `userId` for those lifecycle calls:

- `core/openclaw/src/context-engine/types.ts`
- `core/openclaw/src/agents/pi-embedded-runner/run/attempt.context-engine-helpers.ts`

The host calls look like:

- bootstrap: `sessionId`, `sessionKey`, `sessionFile`
- ingest: `sessionId`, `sessionKey`, `message`
- assemble: `sessionId`, `sessionKey`, `messages`, `tokenBudget`, ...

No `userId`.

## Why status is misleading

LibraVDB status currently counts only prefixes like:

- `turns:`
- `user:`

See `sidecar/server/rpc.go` (`handleStatus`).

That means even if some session-level data were written, status can still report:

- `Turns stored: 0`
- `Memories stored: 0`

So the current status output makes debugging harder because it hides session-only ingest and overstates the failure mode.

## Impact

- Command Center users can install/enable LibraVDB and still get no memory ingestion in normal direct-runtime chat usage
- Even in embedded/runtime-integrated paths, the plugin’s `userId` assumption is incompatible with the current host contract
- `openclaw memory status` does not make the distinction clear

### Plugin-side

1. **Do not require host-supplied `userId` in context-engine args**
   - derive a stable namespace from `sessionKey` when needed
   - or support a sane fallback namespace when `userId` is absent
2. **Make status report session-level ingest too**
   - count `session:*` / raw-session collections separately
   - expose both session-turn count and durable user-memory count
3. **Document the current runtime limitation explicitly**
   - direct runtime vs embedded/runtime-integrated context-engine behavior

### Upstream / host-side

4. **Either** add context-engine lifecycle support to direct runtime
   **or** explicitly document that direct-runtime CC sessions will not ingest into context-engine plugins today
5. If the intended contract requires user identity, extend the host contract to pass a stable user identifier

## Repro notes

Environment I traced:

- OpenClaw: `2026.4.8`
- active session: CC webchat, `Runtime: direct`
- LibraVDB plugin enabled for both:
  - `plugins.slots.memory = libravdb-memory`
  - `plugins.slots.contextEngine = libravdb-memory`
- daemon healthy on Unix socket
- still `0` turns / `0` memories during live conversation

## Practical next step

My recommendation would be to fix the **plugin compatibility issue first**:

- remove the hard dependency on host-provided `userId`
- improve `status()` to reflect actual session ingest

Then separately decide whether direct-runtime support belongs in OpenClaw core or should remain documented as unsupported for context-engine-backed memory plugins.

### PR DESCRIPTION
**Pull request by fuller-stack-dev**
## Summary
- derive the durable memory namespace from `sessionKey` when the host does not supply `userId`, while preserving the existing scoring and gating fields used by retrieval math
- update the memory runtime bridge and CLI namespace operations to resolve the same durable namespace under the latest OpenClaw release contract
- add regression coverage for session-key fallback in context assembly and refresh the math docs to clarify that durable memory can be session-key-derived without changing the formulas

## Testing
- `pnpm run test:ts`
- `pnpm run test:integration`
- `pnpm check` still fails before tests on the pre-existing `scripts/setup.ts` missing declaration for `./child-env.js`

Closes #20.

## Summary by CodeRabbit

* **New Features**
  * CLI: optional session-key for memory flush/export; resolved durable namespace shown.
  * Search: legacy string-query maintained; can load stored-item text via encoded path; status cached for faster probes.

* **Behavior Changes**
  * Durable-namespace resolution (userId → sessionKey → agentId → fallback) now governs ingestion, recall, assembly, and flush; improved ingestion ordering and dedupe.

* **Documentation**
  * Clarified durable-namespace semantics, selection priority, and decay/weighting when userId is absent.

* **Tests**
  * Added unit and integration tests covering durable-namespace, ingestion/assembly, search/read, and export/flush.

### CODE REVIEW

**Comment by fuller-stack-dev** [comment]
Addressed the review set in `849387e`:

- fixed CLI flush/export to use canonical durable namespaces end-to-end, with server support for an explicit `namespace` payload
- tightened namespace resolution by trimming fallback inputs and using session-scoped fallbacks in context-engine paths when `userId`/`sessionKey` are absent
- corrected the memory-runtime legacy/object search return split and encoded both collection/id in result paths
- preserved original host-visible messages in `assemble()` while keeping normalization retrieval-only
- bounded/hash-based `afterTurn` dedupe, fixed last-normalized-message projection handling, and now await raw session inserts before reporting ingest success
- DRYed the shared context namespace args and updated the namespace-order docs
- added regression coverage for resolver trimming, CLI/server namespace behavior, session-scoped fallback, preserved host-visible messages, legacy runtime search shape, encoded result paths, and raw-session-insert failure

Verified with:
- `pnpm run test:ts`
- focused host-flow regression run for the new/changed context-engine cases

The targeted Go RPC test command is still blocked locally by the repo’s documented `sidecar/go.mod` `replace ../../libraVDB` dependency not being present in this checkout.

File: src/context-engine.ts [hunk 1]
```diff
@@ -98,110 +105,91 @@ export function buildContextEngineFactory(
...
+          recallCache,
+          clearElevatedCacheForSession,
+          sessionId,
+          sessionKey,
+          message: {
+            ...normalized,
+            id: `after-turn:${index}`,
+          },
+          skipProjectionRebuild: offset !== normalizedTurnMessages.length - 1,
+        });
```
[the remark this answers is not in the document]
Comment (fuller-stack-dev) [reply src/context-engine.ts]: Addressed in `ac0bfac`. `afterTurn()` now accepts and forwards `userId`, and I added a host-flow regression that verifies post-turn writes stay in `turns:u-explicit` / `user:u-explicit` when an explicit user namespace is available.

File: docs/mathematics-v2.md [hunk 2]
```diff
@@ -200,9 +200,14 @@ $$
...
 config — do not change the decay formula itself.
 
-This makes session context fade fastest, user memory fade more slowly, and
+This makes session context fade fastest, durable namespace memory fade more slowly, and
 global memory remain the most stable.
 
+When the host supplies an explicit `userId`, the durable namespace matches that
+`userId`. When the host does not provide a `userId`, the plugin derives a stable
+durable namespace from the session key so the retrieval math and scope
+weighting stay unchanged even when the host does not expose a separate user principal.
```
[the remark this answers is not in the document]
Comment (fuller-stack-dev) [reply docs/mathematics-v2.md:210]: Addressed in `ac0bfac`. The durable-namespace paragraph now covers all three host-side cases: explicit `userId`, session-key-derived namespace, and the final `session:${sessionId}` fallback when neither identifier is present.

File: src/memory-runtime.ts [hunk 3]
```diff
@@ -140,6 +171,83 @@ function firstString(...values: Array<string | undefined>): string | undefined {
...
+      model: status.embeddingProfile ?? "unknown",
+      ok: status.ok ?? false,
+      message: status.message ?? "ok",
+      turnCount: status.turnCount ?? 0,
+      memoryCount: status.memoryCount ?? 0,
+      gatingThreshold: status.gatingThreshold,
+      abstractiveReady: status.abstractiveReady ?? false,
+      embeddingProfile: status.embeddingProfile ?? "unknown",
+      purpose,
+    };
```
[the remark this answers is not in the document]
Comment (fuller-stack-dev) [reply src/memory-runtime.ts]: Addressed in `ac0bfac`. `readStatus()` now preserves raw sidecar fields by spreading the RPC status payload before applying the normalized fields, and the unit test now asserts that an extra `sessionTurnCount` field survives through `manager.status()`.

**Comment by xDarkicex** [comment]
PR #21 — Technical Implementation & Review

  1. Summary & Reasoning

  - Context: OpenClaw hosts that do not expose a userId principal (e.g., anonymous or internal agent sessions) cannot derive a
  stable durable memory namespace. Memories written under the old "default" fallback collide across sessions. The PR introduces a
  resolveDurableNamespace() priority chain (userId → sessionKey → agentId → fallback → "default") so the plugin can still
  establish a durable namespace from sessionKey when userId is absent.
  - The "Why": The spec (section 17) assumes userId is always supplied. When the host boundary does not provide one, the plugin
  must derive a stable identity from the session key to avoid cross-session memory contamination. os.userInfo() / process.env.USER
   was rejected because it reflects the local OS login identity, not the host conversation identity — mixing local users across
  remote/OpenClaw sessions breaks security isolation and portability.

  2. Issue Tracking

  - Related Issue: Fixes #20
  - Security Impact: No. Namespace collision affects memory isolation, not access control.

  3. Files Touched

  ┌─────────────────────────────────────┬───────────────┬──────────────────────────────────────────────────────────────────────┐
  │                File                 │    Nature     │                             Description                              │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ src/durable-namespace.ts            │ New Feature   │ Centralized namespace resolver: userId → sessionKey → agentId →      │
  │                                     │               │ fallback → "default"                                                 │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │                                     │               │ Replaces direct userId with resolveDurableNamespace() in bootstrap,  │
  │ src/context-engine.ts               │ Refactor      │ ingest, afterTurn, assemble; adds afterTurn hook for post-prompt     │
  │                                     │               │ persist race                                                         │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ src/memory-runtime.ts               │ Refactor      │ Same resolver in search path; adds status caching, readFile,         │
  │                                     │               │ legacy+modern call shapes                                            │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ src/cli.ts                          │ Feature       │ Adds --session-key flag to flush and export commands                 │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ sidecar/server/rpc.go               │ Refactor      │ Adds namespace param to flush_namespace / export_memory; extracts    │
  │                                     │               │ resolveDurableMemoryCollectionPrefix()                               │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ src/types.ts                        │ Refactor      │ userId made optional in ContextBootstrapArgs / ContextIngestArgs /   │
  │                                     │               │ ContextAssembleArgs; new ContextNamespaceArgs base                   │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ docs/gating.md                      │ Documentation │ Renames turns:userId / user:userId → turns:u / user:u; documents     │
  │                                     │               │ namespace resolver priority                                          │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ docs/mathematics-v2.md              │ Documentation │ "durable user memory" → "durable namespace memory" throughout        │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ test/integration/host-flow.test.ts  │ Tests         │ 7 new integration cases covering sessionKey-derived and fallback     │
  │                                     │               │ namespaces                                                           │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ test/unit/durable-namespace.test.ts │ Tests         │ Unit tests for resolveDurableNamespace trimming and priority         │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ test/unit/memory-runtime.test.ts    │ Tests         │ Updated for new resolver path and legacy call shape                  │
  ├─────────────────────────────────────┼───────────────┼──────────────────────────────────────────────────────────────────────┤
  │ sidecar/server/rpc_test.go          │ Tests         │ New namespace-param case for flush_namespace and export_memory       │
  └─────────────────────────────────────┴───────────────┴──────────────────────────────────────────────────────────────────────┘

  4. Implementation Details

  // src/durable-namespace.ts
  export function resolveDurableNamespace(params: {
    userId?: string;
    sessionKey?: string;
    agentId?: string;
    fallback?: string;
  }): string {
    if (userId) return userId;
    if (sessionKey) return `session-key:${sessionKey}`;
    if (agentId) return `agent-id:${agentId}`;
    return fallback ?? "default";
  }

  Technical Decisions

  - Data Structures: resolveDurableNamespace is a pure function — no heap allocation, no mutable state. Threaded through both
  context-engine and memory-runtime paths so both read and write use the same resolved namespace.
  - Concurrency: The afterTurn deduplication cache (afterTurnIngestedKeys as Map<string, number>) uses time-bounded eviction only;
   no locking needed since it is local to the factory closure.

  5. Verification & Testing

  - Unit Tests: pnpm run test:ts — test/unit/durable-namespace.test.ts, test/unit/memory-runtime.test.ts
  - Integration Tests: pnpm run test:integration — test/integration/host-flow.test.ts
  - Sanitizers: (not run in this review pass)
  - Known gap: memory-runtime search resolver is called without sessionId as fallback — see Section 6.

  6. Error Handling & Edge Cases

  Known Limitation — requires pre-merge fix:

  The memory-runtime search() method calls resolveDurableNamespace() without passing sessionId as fallback:

  // src/memory-runtime.ts (PR diff — broken)
  const userId = resolveDurableNamespace({
    userId: firstString(params.userId, params.context?.userId),
    sessionKey: firstString(params.sessionKey, params.context?.sessionKey),
    agentId: firstString(params.agentId, params.context?.agentId, defaults.agentId),
    // fallback is implicitly "default" — wrong for anonymous sessions
  });

  context-engine calls it correctly with fallback: session:${sessionId}`:

  // src/context-engine.ts (PR diff — correct)
  const durableNamespace = resolveDurableNamespace({
    userId, sessionKey,
    fallback: `session:${sessionId}`
  });

  Effect: For an anonymous session with no userId and no sessionKey:
  - context-engine writes to user:session:<sessionId> ✓
  - memory-runtime search looks in user:default ✗ — silent zero-result recall
  - Multiple anonymous sessions collide on user:default ✗

  Required fix: Thread sessionId through MemorySearchParams and pass it as fallback:

  // src/memory-runtime.ts — corrected search call
  const userId = resolveDurableNamespace({
    userId: firstString(params.userId, params.context?.userId),
    sessionKey: firstString(params.sessionKey, params.context?.sessionKey),
    agentId: firstString(params.agentId, params.context?.agentId, defaults.agentId),
    fallback: sessionId ? `session:${sessionId}` : undefined,
  });

  A test should also be added for the search-path fallback case.

  7. Reviewer Checklist

  - resolveDurableNamespace pure function pattern followed — no allocation on hot path
  - sessionId must be threaded into memory-runtime search resolver before merge
  - afterTurn deduplication cache bounded correctly (AFTER_TURN_DEDUPE_TTL_MS, AFTER_TURN_DEDUPE_MAX_ENTRIES)
  - No regression: existing userId-supplied flows are unchanged (userId takes priority)
  - Documentation/Comments updated for namespace resolution priority

  ---
  Once the sessionId-as-fallback` fix is applied and a search-path test is added, this PR is safe to merge. The architecture is
  sound and the 12-file scope is necessary — the resolver contract change touches every call site that writes or reads durable
  memory.

File: src/memory-runtime.ts [hunk 4]
```diff
@@ -51,22 +55,34 @@ function createMemorySearchManager(
...
-        "default",
-      )!;
+      const userId = resolveDurableNamespace({
+        userId: firstString(params.userId, params.context?.userId),
+        sessionKey: firstString(params.sessionKey, params.context?.sessionKey),
+        agentId: firstString(params.agentId, params.context?.agentId, defaults.agentId),
+      });
       const sessionId = firstString(params.sessionId, params.context?.sessionId);
       const k = normalizePositiveInteger(params.k, params.limit, params.topK, cfg.topK, 8);
       const collections = resolveSearchCollections(cfg, userId, sessionId);
```
[the remark this answers is not in the document]
Comment (fuller-stack-dev) [reply src/memory-runtime.ts:89]: Addressed in `d4c874d`. The memory-runtime bridge now passes `fallback: `session:${sessionId}`` into `resolveDurableNamespace()` when no `userId`/`sessionKey`/`agentId` is available, and I added a unit regression that asserts the runtime selects `["session:s-fallback", "user:session:s-fallback", "global"]` for that case.

**Review by xDarkicex — APPROVED** [review]
The sessionId-as-fallbackfix insrc/memory-runtime.ts aligns the search resolver with the context-engine path. The new test
  (memory runtime bridge falls back to session-scoped namespace when no other identity is present) confirms anonymous sessions now
   resolve to user:session:` in both write and search paths.

  Approved for merge.
